### PR TITLE
Security

### PR DESCRIPTION
**Pull request by lukasherz**
please check but don't merge yet, I still have to add the checks if the user has the required role

### CODE REVIEW

File: src/main/java/ch/uzh/ifi/fs22/sel/group15/cozycave/server/controller/AuthenticationController.java [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+    @PostMapping("/register")
+    @ResponseStatus(HttpStatus.CREATED)
+    @ResponseBody
+    public UserGetDto register(@RequestBody UserPostPutDto userPostPutDto) {
+        log.debug("new user registration try: {}", userPostPutDto.toString());
+
+        Location address = null;
+        if (userPostPutDto.getDetails().getAddress() != null) {
+            address = new Location(
+                "home",
```
Comment (matthias-imhof) [inline src/main/java/ch/uzh/ifi/fs22/sel/group15/cozycave/server/controller/AuthenticationController.java:58]: shouldn't the user specify this through userPostPutDto.getDetails().getAddress().getName()? Because of "business" addresses

File: src/main/java/ch/uzh/ifi/fs22/sel/group15/cozycave/server/controller/UserController.java [hunk 2]
```diff
@@ -83,8 +84,8 @@ public UserGetDto updateUser(@PathVariable UUID id, @RequestBody UserPutDto user
     // delete a specific user
     @DeleteMapping("/users/{id}")
     @ResponseStatus(HttpStatus.NO_CONTENT)
-    public void deleteUser(@PathVariable UUID id, @RequestBody UserPutDto userPutDto) {
-        User userInput = UserMapper.INSTANCE.userPutDtoToUser(userPutDto);
+    public void deleteUser(@PathVariable UUID id, @RequestBody UserPostPutDto userPostPutDto) {
+        User userInput = UserMapper.INSTANCE.userPostPutDtoToUser(userPostPutDto);
```
Comment (matthias-imhof) [inline src/main/java/ch/uzh/ifi/fs22/sel/group15/cozycave/server/controller/UserController.java:89]: User user = userService.findUserID(id)
            .orElseThrow(
                () -> new ResponseStatusException(HttpStatus.NOT_FOUND, "User couldn't be found with that user ID."));

File: src/main/java/ch/uzh/ifi/fs22/sel/group15/cozycave/server/service/UserService.java [hunk 3]
```diff
@@ -26,37 +28,78 @@
...
         return this.userRepository.findAll();
     }
 
-    public @NotNull User createUser(User newUser) {
+    public @NotNull User createUser(User newUser, User createdBy) {
+        log.debug("creating user {}", newUser);
+
+        newUser.setId(null);
         newUser.setCreationDate(new Date());
-        newUser.getAuthenticationData().setSalt(Utils.generateSalt(16));
```
Comment (matthias-imhof) [inline src/main/java/ch/uzh/ifi/fs22/sel/group15/cozycave/server/service/UserService.java:40]: Shouldn't we use a static class variable here?     Utils.SALT_LENGTH

File: src/main/java/ch/uzh/ifi/fs22/sel/group15/cozycave/server/constant/Role.java [hunk 4]
```diff
@@ -1,12 +1,16 @@
...
 import org.jetbrains.annotations.Nullable;
+import org.springframework.security.core.authority.SimpleGrantedAuthority;
 
 public enum Role {
     LANDLORD(1),
     STUDENT(2),
     TEAM(100),
-    ADMIN(999);
+    ADMIN(999),
+    INTERNAL(1000);
```
[the remark this answers is not in the document]
Comment (lukasherz) [reply src/main/java/ch/uzh/ifi/fs22/sel/group15/cozycave/server/constant/Role.java:13]: if we need a test user with maximum rights. this would just be for testing internal usage etc. its like an super user

File: src/main/java/ch/uzh/ifi/fs22/sel/group15/cozycave/server/controller/AuthenticationController.java (hunk 1 above)
Comment (lukasherz) [reply src/main/java/ch/uzh/ifi/fs22/sel/group15/cozycave/server/controller/AuthenticationController.java:58]: should this then be required?

File: src/main/java/ch/uzh/ifi/fs22/sel/group15/cozycave/server/service/UserService.java (hunk 3 above)
Comment (lukasherz) [reply src/main/java/ch/uzh/ifi/fs22/sel/group15/cozycave/server/service/UserService.java:40]: as you can see, this was already removed and changed

File: src/main/java/ch/uzh/ifi/fs22/sel/group15/cozycave/server/constant/Role.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (matthias-imhof) [reply src/main/java/ch/uzh/ifi/fs22/sel/group15/cozycave/server/constant/Role.java:13]: I see makes sense, but we need to keep in mind to make it more secure in the future to not easily allow to create a user with god privillegages.

File: src/main/java/ch/uzh/ifi/fs22/sel/group15/cozycave/server/controller/AuthenticationController.java (hunk 1 above)
Comment (matthias-imhof) [reply src/main/java/ch/uzh/ifi/fs22/sel/group15/cozycave/server/controller/AuthenticationController.java]: yes I guess, but we can do this later as well.

**Review by matthias-imhof — APPROVED** [review]

File: src/main/java/ch/uzh/ifi/fs22/sel/group15/cozycave/server/constant/Role.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (lukasherz) [reply src/main/java/ch/uzh/ifi/fs22/sel/group15/cozycave/server/constant/Role.java:13]: yeah of course. ill later would have to implement the role based authorization. this would be applicable in M4
